be/src/runtime/batch_write/txn_state_cache.cpp

```diff
@@ -0,0 +1,467 @@
+    TRACE_BATCH_WRITE << "submit poll task, txn_id: " << task.txn_id << ", db: " << task.db << ", tbl: " << task.tbl
+                      << ", delay_ms: " << delay_ms;
+}
+
+void TxnStatePoller::stop() {
+    {
+        std::unique_lock<bthread::Mutex> lock(_mutex);
+        if (_stopped) {
+            return;
+        }
+        _stopped = true;
+        _cv.notify_all();
+    }
+    if (_schedule_thread && _schedule_thread->joinable()) {
+        _schedule_thread->join();
+    }
+}
+
+void TxnStatePoller::_schedule_func() {
+    std::vector<TxnStatePollTask> poll_tasks;
+    std::unique_lock<bthread::Mutex> lock(_mutex);
+    _is_scheduling = true;
+    while (!_stopped) {
+        int64_t current_ts = get_current_ms();
+        auto it = _pending_tasks.begin();
+        while (it != _pending_tasks.end()) {
+            if (it->first <= current_ts) {
+                _pending_txn_ids.erase(it->second.txn_id);
+                poll_tasks.emplace_back(it->second);
+                it = _pending_tasks.erase(it);
+            } else {
+                break;
+            }
+        }
+        if (!poll_tasks.empty()) {
+            lock.unlock();
+            _schedule_poll_tasks(poll_tasks);
+            poll_tasks.clear();
+            lock.lock();
+        }
+        if (_stopped) {
+            break;
+        }
+        if (_pending_tasks.empty()) {
+            _cv.wait(lock);
+        } else {
+            // at least wait 50 ms to avoid busy loop
+            int64_t wait_time_ms = std::max((int64_t)50, _pending_tasks.begin()->first - get_current_ms());
+            _cv.wait_for(lock, wait_time_ms * 1000);
+        }
+    }
+    _is_scheduling = false;
+}
+
+void TxnStatePoller::_schedule_poll_tasks(const std::vector<TxnStatePollTask>& poll_tasks) {
+    for (const auto& task : poll_tasks) {
+        // check current state of the txn, if it's not in cache or is final, skip poll
+        auto current_state = _txn_state_cache->get_state(task.txn_id);
+        if (!current_state.ok()) {
+            TRACE_BATCH_WRITE << "skip poll task because fail to get txn state, txn_id: " << task.txn_id
+                              << ", db: " << task.db << ", tbl: " << task.tbl << ", error: " << current_state.status();
+            continue;
+        }
+        TTransactionStatus::type txn_status = current_state.value().txn_status;
+        if (txn_status == TTransactionStatus::VISIBLE || txn_status == TTransactionStatus::ABORTED ||
+            txn_status == TTransactionStatus::UNKNOWN) {
+            TRACE_BATCH_WRITE << "skip poll task because txn state is final, txn_id: " << task.txn_id
+                              << ", db: " << task.db << ", tbl: " << task.tbl << ", state status: " << txn_status
+                              << ", reason: " << current_state.value().reason;
+            continue;
+        }
+        Status status = _poll_token->submit_func([this, task] { _execute_poll(task); }, ThreadPool::HIGH_PRIORITY);
```

Comment:
is it possible to merge all the txn ids into a single FE RPC, batch pull the txn status, save the rpc cost?

Reply:
it should be. I plan to optimize it later. `getLoadStatus` only support one txn in one rpc. I need to add a new one